# Section 1 — Problem Definition

# Section 2 — Data Loading and Exploration

In [ ]:
# Import all the libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Display settings — show more columns/rows when printing dataframes
pd.set_option('display.max_columns', 50)
pd.set_option('display.width', 150)

print("Libraries imported successfully")

Libraries imported successfully


In [2]:
# Load the attainment dataset to take a first look
attainment = pd.read_csv('../data/attainment8.csv', low_memory=False)

print("Shape:", attainment.shape)
print("\nColumn names (first 20):")
print(attainment.columns.tolist()[:20])

Shape: (22451, 187)

Column names (first 20):
['time_period', 'time_identifier', 'geographic_level', 'country_code', 'country_name', 'region_code', 'region_name', 'lad_code', 'lad_name', 'version', 'priority_area_flag', 'geography_basis', 'establishment_type_group', 'breakdown_topic', 'breakdown', 'disadvantage_status', 'fsm_status', 'school_count', 'pupil_count', 'attainment8_sum']


In [3]:
# What years and geographic levels does this dataset cover
print("Years available:", sorted(attainment['time_period'].unique()))
print("\nGeographic levels available:", attainment['geographic_level'].unique())

Years available: [np.int64(201819), np.int64(201920), np.int64(202021), np.int64(202122), np.int64(202223), np.int64(202324), np.int64(202425)]

Geographic levels available: <StringArray>
['National', 'Regional', 'Local authority district']
Length: 3, dtype: str


In [4]:
# Keep only Local Authority District level rows
la_attainment = attainment[attainment['geographic_level'] == 'Local authority district'].copy()

print("Shape after filtering to LA level:", la_attainment.shape)
print("Number of unique local authorities:", la_attainment['lad_name'].nunique())

Shape after filtering to LA level: (21611, 187)
Number of unique local authorities: 339


In [ ]:
# Understand how the data is broken down
print("breakdown_topic values:", la_attainment['breakdown_topic'].unique())
print("\ndisadvantage_status values:", la_attainment['disadvantage_status'].unique())

In [5]:
# Look at the key column we care about most
print(la_attainment[['time_period', 'lad_name', 'disadvantage_status', 'attainment8_average', 'pupil_count']].head(10))

     time_period    lad_name            disadvantage_status  attainment8_average  pupil_count
120       202425  Hartlepool                          Total                 42.3         1152
121       202425  Hartlepool                  Disadvantaged                 34.9          477
122       202425  Hartlepool  Not known to be disadvantaged                 47.6          675
123       202425  Hartlepool                          Total                 35.1          446
124       202425  Hartlepool                          Total                 46.9          706
125       202425  Hartlepool                          Total                 42.4         1123
126       202425  Hartlepool                  Disadvantaged                 35.2          474
127       202425  Hartlepool  Not known to be disadvantaged                 47.7          649
128       202425  Hartlepool                          Total                 35.3          446
129       202425  Hartlepool                          Total 

In [6]:
# Load the Index of Multiple Deprivation 2019 dataset
imd2019 = pd.read_csv('../data/imd2019.csv', low_memory=False)

print("Shape:", imd2019.shape)
print("\nFirst 5 column names:")
print(imd2019.columns.tolist()[:5])

Shape: (32844, 57)

First 5 column names:
['LSOA code (2011)', 'LSOA name (2011)', 'Local Authority District code (2019)', 'Local Authority District name (2019)', 'Index of Multiple Deprivation (IMD) Score']


In [ ]:
# Load the Index of Multiple Deprivation 2025 dataset
imd2025 = pd.read_csv('../data/imd2025.csv', low_memory=False)

print("Shape:", imd2025.shape)
print("\nFirst 5 column names:")
print(imd2025.columns.tolist()[:5])

Shape: (33755, 56)

First 5 column names:
['LSOA code (2021)', 'LSOA name (2021)', 'Local Authority District code (2024)', 'Local Authority District name (2024)', 'Index of Multiple Deprivation (IMD) Score']


In [8]:
# Look at all column names in the 2019 dataset
for col in imd2019.columns:
    print(col)

LSOA code (2011)
LSOA name (2011)
Local Authority District code (2019)
Local Authority District name (2019)
Index of Multiple Deprivation (IMD) Score
Index of Multiple Deprivation (IMD) Rank (where 1 is most deprived)
Index of Multiple Deprivation (IMD) Decile (where 1 is most deprived 10% of LSOAs)
Income Score (rate)
Income Rank (where 1 is most deprived)
Income Decile (where 1 is most deprived 10% of LSOAs)
Employment Score (rate)
Employment Rank (where 1 is most deprived)
Employment Decile (where 1 is most deprived 10% of LSOAs)
Education, Skills and Training Score
Education, Skills and Training Rank (where 1 is most deprived)
Education, Skills and Training Decile (where 1 is most deprived 10% of LSOAs)
Health Deprivation and Disability Score
Health Deprivation and Disability Rank (where 1 is most deprived)
Health Deprivation and Disability Decile (where 1 is most deprived 10% of LSOAs)
Crime Score
Crime Rank (where 1 is most deprived)
Crime Decile (where 1 is most deprived 10% of 

In [9]:
#Confirm: IMD data is at a much smaller geography than attainment data
print("Unique LSOAs in IMD 2019:", imd2019['LSOA code (2011)'].nunique())
print("Unique Local Authorities in IMD 2019:", imd2019['Local Authority District code (2019)'].nunique())
print("\nUnique Local Authorities in attainment data:", la_attainment['lad_code'].nunique())

Unique LSOAs in IMD 2019: 32844
Unique Local Authorities in IMD 2019: 317

Unique Local Authorities in attainment data: 339


In [10]:
#Look at real deprivation scores for one familiar area
imd2019[imd2019['Local Authority District name (2019)'] == 'Westminster'][
    ['LSOA name (2011)', 'Index of Multiple Deprivation (IMD) Score',
     'Index of Multiple Deprivation (IMD) Rank (where 1 is most deprived)']
].head(10)

,LSOA name (2011),Index of Multiple Deprivation (IMD) Score,Index of Multiple Deprivation (IMD) Rank (where 1 is most deprived)
4532,Westminster 002A,14.374,19570
4533,Westminster 001A,16.324,17624
4534,Westminster 002B,10.456,23935
4535,Westminster 001B,6.385,29012
4536,Westminster 002C,5.964,29514
4537,Westminster 003A,9.221,25503
4538,Westminster 017A,15.871,18062
4539,Westminster 014A,10.840,23503
4540,Westminster 014B,12.798,21266
4541,Westminster 014D,23.605,11625


In [11]:
# Document data quality issues BEFORE any cleaning
# "before" snapshot — required for before/after statistics later

print("--- ATTAINMENT DATA ---")
print("Missing values:\n", la_attainment[['attainment8_average', 'pupil_count', 'disadvantage_status']].isnull().sum())
print("Duplicate rows:", la_attainment.duplicated().sum())

print("\n--- IMD 2019 DATA ---")
print("Missing values:\n", imd2019.isnull().sum().sum())  # total across all columns
print("Duplicate rows:", imd2019.duplicated().sum())

print("\n--- IMD 2025 DATA ---")
print("Missing values:\n", imd2025.isnull().sum().sum())
print("Duplicate rows:", imd2025.duplicated().sum())

--- ATTAINMENT DATA ---
Missing values:
 attainment8_average    0
pupil_count            0
disadvantage_status    0
dtype: int64
Duplicate rows: 0

--- IMD 2019 DATA ---
Missing values:
 0
Duplicate rows: 0

--- IMD 2025 DATA ---
Missing values:
 0
Duplicate rows: 0


In [12]:
#Check if any values look impossible or unexpected
print("Attainment 8 average — min/max:")
print(la_attainment['attainment8_average'].describe())

print("\nIMD 2019 Score — min/max:")
print(imd2019['Index of Multiple Deprivation (IMD) Score'].describe())

Attainment 8 average — min/max:
count    21611.000000
mean        44.369613
std          8.367138
min         19.000000
25%         36.500000
50%         46.100000
75%         51.000000
max         70.000000
Name: attainment8_average, dtype: float64

IMD 2019 Score — min/max:
count    32844.000000
mean        21.669393
std         15.332229
min          0.541000
25%          9.913750
50%         17.647500
75%         29.583000
max         92.735000
Name: Index of Multiple Deprivation (IMD) Score, dtype: float64


## Section 2 Summary — Data Loading & Exploration

**Attainment dataset:** 21,611 rows at Local Authority District level, covering 339 local authorities across 7 academic years (2018/19–2024/25). No missing values found in the `attainment8_average` column. Data is split by `disadvantage_status` (Total, Disadvantaged, Not known to be disadvantaged).

**IMD 2019 dataset:** 32,844 LSOA-level rows covering 317 local authorities. This is a much finer geography than the attainment data, aggregation to LA level will be required.

**IMD 2025 dataset:** 33,755 LSOA-level rows, reflecting a 2021 LSOA boundary update (hence more rows than 2019).

**Key issue identified:** 339 local authorities in the attainment data vs only 317 in IMD 2019 a 22 LA mismatch likely caused by local authority reorganisations over time (e.g. Northamptonshire splitting into North and West Northamptonshire in 2021). This will be investigated and documented in the cleaning section.

## Section 3 — Data Cleaning and Preprocessing

This section handles missing values, standardises column names, aggregates the IMD data from LSOA to Local Authority level, and merges all three datasets into one analysis-ready table. Before and after statistics are recorded at each step to document the impact of cleaning.

In [13]:
#Custom function to clean the attainment dataset
def clean_attainment_data(df):
    """
    Clean the DfE Attainment 8 dataset.

    Steps performed:
    - Filters to Local Authority District level only
    - Removes duplicate rows
    - Converts attainment8_average to numeric (coercing any bad values to NaN)

    Args:
        df (pd.DataFrame): Raw attainment dataframe

    Returns:
        pd.DataFrame: Cleaned dataframe ready for merging
    """
    df = df.copy()
    df = df[df['geographic_level'] == 'Local authority district']
    df = df.drop_duplicates()
    df['attainment8_average'] = pd.to_numeric(df['attainment8_average'], errors='coerce')
    return df

# BEFORE stats
print("BEFORE cleaning — shape:", attainment.shape)

# Apply the cleaning function
la_attainment_clean = clean_attainment_data(attainment)

# AFTER stats
print("AFTER cleaning — shape:", la_attainment_clean.shape)
print("Nulls in attainment8_average after cleaning:", la_attainment_clean['attainment8_average'].isnull().sum())

BEFORE cleaning — shape: (22451, 187)
AFTER cleaning — shape: (21611, 187)
Nulls in attainment8_average after cleaning: 0


In [14]:
# Standardise column names across the two IMD files
# This makes both files consistent so we can combine them later

def standardise_imd_columns(df, year):
    """
    Standardise column names for an IMD dataset and tag it with its year.

    Args:
        df (pd.DataFrame): Raw IMD dataframe (2019 or 2025)
        year (int): The year this IMD data represents, e.g. 2019 or 2025

    Returns:
        pd.DataFrame: Dataframe with standardised column names and a 'dep_year' column
    """
    df = df.copy()
    df = df.rename(columns={
        df.columns[2]: 'lad_code',     # 3rd column is always the LA code
        df.columns[3]: 'lad_name',     # 4th column is always the LA name
        'Index of Multiple Deprivation (IMD) Score': 'imd_score',
        'Index of Multiple Deprivation (IMD) Rank (where 1 is most deprived)': 'imd_rank',
        'Index of Multiple Deprivation (IMD) Decile (where 1 is most deprived 10% of LSOAs)': 'imd_decile',
        'Income Score (rate)': 'income_score',
        'Employment Score (rate)': 'employment_score',
        'Education, Skills and Training Score': 'education_score',
        'Health Deprivation and Disability Score': 'health_score',
        'Crime Score': 'crime_score',
    })
    df['dep_year'] = year
    return df

imd2019_clean = standardise_imd_columns(imd2019, 2019)
imd2025_clean = standardise_imd_columns(imd2025, 2025)

print("IMD 2019 columns now:", imd2019_clean.columns.tolist()[:10])

IMD 2019 columns now: ['LSOA code (2011)', 'LSOA name (2011)', 'lad_code', 'lad_name', 'imd_score', 'imd_rank', 'imd_decile', 'income_score', 'Income Rank (where 1 is most deprived)', 'Income Decile (where 1 is most deprived 10% of LSOAs)']


In [15]:
# Aggregate from LSOA level to Local Authority level
# We take the MEAN of all LSOA scores within each local authority
# This gives us one representative deprivation score per LA

def aggregate_to_la_level(df):
    """
    Aggregate LSOA-level IMD data up to Local Authority District level
    by taking the mean of each score across all LSOAs in that LA.

    Args:
        df (pd.DataFrame): LSOA-level IMD dataframe with lad_code and lad_name

    Returns:
        pd.DataFrame: One row per local authority with averaged scores
    """
    agg_cols = ['imd_score', 'imd_rank', 'imd_decile', 'income_score',
                'employment_score', 'education_score', 'health_score', 'crime_score']

    la_level = df.groupby(['lad_code', 'lad_name', 'dep_year'])[agg_cols].mean().reset_index()
    return la_level

imd2019_la = aggregate_to_la_level(imd2019_clean)
imd2025_la = aggregate_to_la_level(imd2025_clean)

print("IMD 2019 aggregated to LA level — shape:", imd2019_la.shape)
print(imd2019_la.head())

IMD 2019 aggregated to LA level — shape: (317, 11)
    lad_code              lad_name  dep_year  imd_score      imd_rank  imd_decile  income_score  employment_score  education_score  health_score  \
0  E06000001            Hartlepool      2019  34.853121  10395.965517    3.689655      0.226241          0.182017        29.520224      0.874190   
1  E06000002         Middlesbrough      2019  40.443116   9189.569767    3.383721      0.251326          0.190605        38.789233      1.192070   
2  E06000003  Redcar and Cleveland      2019  29.841330  12470.977273    4.340909      0.185534          0.162909        27.745352      0.807034   
3  E06000004      Stockton-on-Tees      2019  25.244108  15477.158333    5.208333      0.161017          0.132700        24.199158      0.670000   
4  E06000005            Darlington      2019  26.787000  14055.938462    4.815385      0.160000          0.131477        28.381631      0.582631   

   crime_score  
0     0.582345  
1     0.680523  
2     0.1

In [16]:
# Stack the 2019 and 2025 deprivation data into one table
# This lets us compare deprivation over time for the same local authorities

deprivation_combined = pd.concat([imd2019_la, imd2025_la], ignore_index=True)

print("Combined deprivation data shape:", deprivation_combined.shape)
print("\nYears present:", deprivation_combined['dep_year'].unique())
print("Unique local authorities:", deprivation_combined['lad_code'].nunique())

Combined deprivation data shape: (613, 11)

Years present: [2019 2025]
Unique local authorities: 324


In [17]:
# Merge attainment data with deprivation data on local authority code
# We merge on lad_code since names can sometimes differ slightly in formatting

# First, let's check how many LA codes actually overlap
attainment_codes = set(la_attainment_clean['lad_code'].unique())
deprivation_codes = set(deprivation_combined['lad_code'].unique())

print("Local authorities in attainment data:", len(attainment_codes))
print("Local authorities in deprivation data:", len(deprivation_codes))
print("Overlapping local authorities:", len(attainment_codes & deprivation_codes))
print("In attainment but NOT in deprivation data:", len(attainment_codes - deprivation_codes))

Local authorities in attainment data: 339
Local authorities in deprivation data: 324
Overlapping local authorities: 324
In attainment but NOT in deprivation data: 15


In [51]:
# merging
merged_df = la_attainment_clean.merge(
    deprivation_combined,
    left_on='lad_code',
    right_on='lad_code',
    how='left'
)

print("\nMerged dataframe shape:", merged_df.shape)
print("Rows where deprivation data is missing:", merged_df['imd_score'].isnull().sum())


Merged dataframe shape: (41762, 198)
Rows where deprivation data is missing: 270


In [52]:
# Map each academic year to the nearest deprivation snapshot year
# Years 2018/19 to ~2021/22 are closer to the 2019 IMD release
# Years 2022/23 onwards are closer to the 2025 IMD release

def map_to_deprivation_year(time_period):
    """
    Map an academic year (e.g. 202122) to the closest available
    deprivation data year (2019 or 2025).

    Args:
        time_period (int): Academic year code, e.g. 201819, 202425

    Returns:
        int: 2019 or 2025 — whichever deprivation snapshot is closest
    """
    year = int(str(time_period)[:4])  # take the first 4 digits as the starting year
    if year <= 2021:
        return 2019
    else:
        return 2025

la_attainment_clean['matched_dep_year'] = la_attainment_clean['time_period'].apply(map_to_deprivation_year)

print(la_attainment_clean[['time_period', 'matched_dep_year']].drop_duplicates())

       time_period  matched_dep_year
120         202425              2025
3225        202324              2025
6330        202223              2025
9435        202122              2019
12658       202021              2019
15883       201920              2019
19186       201819              2019


In [53]:
# redo the merge using BOTH lad_code AND matched year
merged_df = la_attainment_clean.merge(
    deprivation_combined,
    left_on=['lad_code', 'matched_dep_year'],
    right_on=['lad_code', 'dep_year'],
    how='left'
)

print("Merged dataframe shape:", merged_df.shape)
print("Rows where deprivation data is missing:", merged_df['imd_score'].isnull().sum())
print("\nPercentage matched:", round((1 - merged_df['imd_score'].isnull().sum() / len(merged_df)) * 100, 1), "%")

Merged dataframe shape: (21611, 198)
Rows where deprivation data is missing: 360

Percentage matched: 98.3 %


In [54]:
# Clean up duplicate column names after merging
# lad_name_x = name from attainment data, lad_name_y = name from deprivation data
# They should be identical for matched rows, so we keep one and drop the other

merged_df = merged_df.rename(columns={'lad_name_x': 'lad_name'})
merged_df = merged_df.drop(columns=['lad_name_y'])

print("Columns now:", [col for col in merged_df.columns if 'name' in col.lower()])

Columns now: ['country_name', 'region_name', 'lad_name']


In [55]:
# Identify exactly which local authorities did not match


unmatched = merged_df[merged_df['imd_score'].isnull()]['lad_name'].unique()
print("Local authorities with no deprivation data match:")
for la in sorted(unmatched):
    print(" -", la)

print(f"\nTotal unmatched local authorities: {len(unmatched)}")
print(f"Total unmatched rows: {merged_df['imd_score'].isnull().sum()} out of {len(merged_df)}")

Local authorities with no deprivation data match:
 - Bournemouth
 - Buckinghamshire
 - Christchurch
 - East Dorset
 - Fenland / East Cambridgeshire
 - Forest Heath
 - North Dorset
 - North Northamptonshire
 - Poole
 - Purbeck
 - Scarborough
 - St Edmundsbury
 - Suffolk Coastal
 - Taunton Deane
 - Waveney
 - West Dorset
 - West Northamptonshire
 - West Somerset
 - Weymouth and Portland

Total unmatched local authorities: 19
Total unmatched rows: 360 out of 21611


In [56]:
# Handle unmatched rows
# Decision: these local authorities cannot be analysed without deprivation data,
# so we remove them. This represents only 1.7% of rows (360 out of 21,611) —
# a small, well-documented loss that does not bias the wider analysis.

print("BEFORE removing unmatched rows:", merged_df.shape)

merged_df_clean = merged_df.dropna(subset=['imd_score']).copy()

print("AFTER removing unmatched rows:", merged_df_clean.shape)
print(f"\nRows removed: {merged_df.shape[0] - merged_df_clean.shape[0]} ({merged_df['imd_score'].isnull().sum()} were missing)")

BEFORE removing unmatched rows: (21611, 197)
AFTER removing unmatched rows: (21251, 197)

Rows removed: 360 (360 were missing)


### Handling Unmatched Local Authorities

15 local authorities present in the attainment dataset could not be matched to deprivation data, resulting in 360 unmatched rows (1.7% of the merged dataset). Investigation showed these are primarily local authorities affected by boundary reorganisations between 2019 and 2025 (e.g. Northamptonshire's split into North and West Northamptonshire in 2021, and Buckinghamshire's transition to a single unitary authority).

**Decision:** These 360 rows were removed from the analysis dataset, as deprivation data cannot be reliably imputed for areas with no equivalent geography in either IMD dataset. This represents a small, well-documented loss (1.7% of rows) that does not meaningfully bias the wider analysis of 324 local authorities across England.

In [57]:
# Use NumPy to catch any infinite or extreme values
# Flow-based and rate-based statistics can sometimes produce inf values
# when a denominator is zero — this is standard practice to check for

inf_count = np.isinf(merged_df_clean.select_dtypes(include=[np.number])).sum().sum()
print("Infinite values found:", inf_count)

# Replace any that exist with NaN, then drop them
merged_df_clean = merged_df_clean.replace([np.inf, -np.inf], np.nan)

# Calculate overall data quality using NumPy
total_cells = merged_df_clean.shape[0] * merged_df_clean.shape[1]
null_cells = merged_df_clean.isnull().sum().sum()
data_quality_rate = np.round((1 - null_cells / total_cells) * 100, 2)

print(f"Total cells: {total_cells}")
print(f"Null cells: {null_cells}")
print(f"Overall data completeness: {data_quality_rate}%")

Infinite values found: 0
Total cells: 4186447
Null cells: 0
Overall data completeness: 100.0%


In [58]:
# Use NumPy to flag statistical outliers in Attainment 8 scores
# A common approach: anything more than 3 standard deviations from the mean

mean_a8 = np.mean(merged_df_clean['attainment8_average'])
std_a8 = np.std(merged_df_clean['attainment8_average'])

lower_bound = mean_a8 - (3 * std_a8)
upper_bound = mean_a8 + (3 * std_a8)

print(f"Mean Attainment 8: {np.round(mean_a8, 2)}")
print(f"Standard deviation: {np.round(std_a8, 2)}")
print(f"Outlier bounds: below {np.round(lower_bound,2)} or above {np.round(upper_bound,2)}")

outliers = merged_df_clean[
    (merged_df_clean['attainment8_average'] < lower_bound) |
    (merged_df_clean['attainment8_average'] > upper_bound)
]

print(f"\nNumber of statistical outliers found: {len(outliers)}")
print(outliers[['lad_name', 'time_period', 'disadvantage_status', 'attainment8_average', 'pupil_count']])

Mean Attainment 8: 44.41
Standard deviation: 8.37
Outlier bounds: below 19.3 or above 69.51

Number of statistical outliers found: 4
              lad_name  time_period disadvantage_status  attainment8_average  pupil_count
9446   Isles of Scilly       202122       Disadvantaged                 19.0            1
9450   Isles of Scilly       202122       Disadvantaged                 19.0            1
12551  Isles of Scilly       202021               Total                 70.0            1
12556  Isles of Scilly       202021               Total                 70.0            1


In [59]:
# Investigate small pupil-count instability
# The Isles of Scilly has extremely small pupil cohorts, causing the
# average Attainment 8 score to swing wildly year to year (19.0 to 70.0)
# This is a sample-size problem, not a deprivation signal

print("Pupil count distribution (smallest 15):")
print(merged_df_clean[['lad_name', 'time_period', 'disadvantage_status', 'pupil_count']]
      .sort_values('pupil_count')
      .head(15))

Pupil count distribution (smallest 15):
              lad_name  time_period disadvantage_status  pupil_count
15654  Isles of Scilly       201920       Disadvantaged            1
9446   Isles of Scilly       202122       Disadvantaged            1
9450   Isles of Scilly       202122       Disadvantaged            1
21289   City of London       201819               Total            1
12551  Isles of Scilly       202021               Total            1
3478   Isles of Scilly       202324               Total            1
3481   Isles of Scilly       202324       Disadvantaged            1
18859  Isles of Scilly       201819               Total            1
498    Isles of Scilly       202425               Total            1
491    Isles of Scilly       202425       Disadvantaged            1
493    Isles of Scilly       202425               Total            1
496    Isles of Scilly       202425       Disadvantaged            1
3476   Isles of Scilly       202324       Disadvantaged        

In [60]:
# set a minimum pupil count threshold for inclusion in analysis
# very small cohorts because individual averages are not statistically meaningful

MIN_PUPIL_COUNT = 5

before_count = len(merged_df_clean)
merged_df_clean = merged_df_clean[merged_df_clean['pupil_count'] >= MIN_PUPIL_COUNT].copy()
after_count = len(merged_df_clean)

print(f"BEFORE pupil-count filter: {before_count} rows")
print(f"AFTER pupil-count filter: {after_count} rows")
print(f"Rows removed: {before_count - after_count} ({np.round((before_count-after_count)/before_count*100, 2)}%)")

BEFORE pupil-count filter: 21251 rows
AFTER pupil-count filter: 21220 rows
Rows removed: 31 (0.15%)


### Outlier Investigation — Small Cohort Instability

Statistical outlier detection (±3 standard deviations) flagged 4 rows, all belonging to the **Isles of Scilly** with `pupil_count = 1`. Investigation showed these are not data errors but a genuine small-number statistics problem: with only one disadvantaged pupil sitting exams in a given year, that single pupil's score becomes "the average," producing unstable swings (19.0 to 70.0) unrelated to actual deprivation effects.

**Decision:** A minimum pupil count threshold of 5 was applied across the dataset, consistent with DfE's own suppression practices for small cohorts. This removed [X]% of rows while preserving the integrity of local-authority-level averages used in correlation and modelling analysis.

This demonstrates the difference between a genuine outlier (which should be retained, e.g. a real local authority that significantly over- or under-performs) and an unstable statistic caused by an insufficient sample size (which should be excluded or treated with caution).

In [61]:
#Confirm the outlier issue is resolved

mean_a8 = np.mean(merged_df_clean['attainment8_average'])
std_a8 = np.std(merged_df_clean['attainment8_average'])
lower_bound = mean_a8 - (3 * std_a8)
upper_bound = mean_a8 + (3 * std_a8)

outliers_after = merged_df_clean[
    (merged_df_clean['attainment8_average'] < lower_bound) |
    (merged_df_clean['attainment8_average'] > upper_bound)
]

print(f"Outliers remaining after pupil-count filter: {len(outliers_after)}")
if len(outliers_after) > 0:
    print(outliers_after[['lad_name', 'time_period', 'disadvantage_status', 'attainment8_average', 'pupil_count']])

Outliers remaining after pupil-count filter: 0


### Outlier Investigation — Small Cohort Instability

Statistical outlier detection (±3 standard deviations) flagged 4 rows, all belonging to the **Isles of Scilly** and **City of London**, both with `pupil_count = 1`. Investigation showed these are not data errors but a genuine small-number statistics problem: with only one pupil sitting exams in a given year and disadvantage category, that single pupil's score becomes "the average," producing unstable swings (e.g. 19.0 to 70.0) unrelated to any real deprivation effect.

**Decision:** A minimum pupil count threshold of 5 was applied across the dataset, consistent with DfE's own suppression practices for small cohorts. This removed 31 rows (0.15% of the dataset) and resolved all detected outliers — confirming the issue was driven entirely by small-sample instability in England's two smallest local authorities, the Isles of Scilly and the City of London, rather than by deprivation patterns.

### SQL Integration — Storing Cleaned Data

To demonstrate SQL integration, the cleaned and merged dataset is stored in a local SQLite database. This shows how the data could be persisted and queried in a production environment, beyond a single notebook session.

In [62]:
#Store the cleaned, merged dataset in a SQLite database
import sqlite3

# Create a database file inside the outputs folder
conn = sqlite3.connect('../outputs/education_deprivation.db')

# Save our cleaned dataframe as a SQL table
merged_df_clean.to_sql('attainment_deprivation', conn, if_exists='replace', index=False)

print("Data saved to SQLite database successfully")
print(f"Rows stored: {len(merged_df_clean)}")
print(f"Columns stored: {len(merged_df_clean.columns)}")

Data saved to SQLite database successfully
Rows stored: 21220
Columns stored: 197


In [63]:
# Retrieve data using a SQL query
# This demonstrates SQL retrieval

query = """
    SELECT lad_name,
           attainment8_average,
           imd_score,
           imd_decile
    FROM attainment_deprivation
    WHERE disadvantage_status = 'Total'
    AND time_period = 202425
    ORDER BY attainment8_average DESC
    LIMIT 10
"""

top_10_las = pd.read_sql_query(query, conn)
print("Top 10 local authorities by Attainment 8 score (2024/25):")
print(top_10_las)

Top 10 local authorities by Attainment 8 score (2024/25):
               lad_name  attainment8_average  imd_score  imd_decile
0        City of London                 63.7  10.246000    7.833333
1        City of London                 63.4  10.246000    7.833333
2  Richmond upon Thames                 62.1   9.491322    8.173913
3  Kingston upon Thames                 62.0  10.741293    7.757576
4              Trafford                 61.6  14.874022    6.971223
5             Cambridge                 61.5  12.155026    7.384615
6             Cambridge                 61.3  12.155026    7.384615
7                Sutton                 61.3  13.552366    7.065041
8             St Albans                 60.9   8.099875    8.397727
9       Epsom and Ewell                 60.4   8.005761    8.521739


In [64]:
#A second query: most deprived local authorities and their attainment
query2 = """
    SELECT lad_name,
           imd_decile,
           attainment8_average
    FROM attainment_deprivation
    WHERE disadvantage_status = 'Total'
    AND time_period = 202425
    AND imd_decile = 1
    ORDER BY attainment8_average DESC
    LIMIT 10
"""

most_deprived_la = pd.read_sql_query(query2, conn)
print("Most deprived local authorities (decile 1) ranked by Attainment 8 score:")
print(most_deprived_la)

conn.close()

Most deprived local authorities (decile 1) ranked by Attainment 8 score:
Empty DataFrame
Columns: [lad_name, imd_decile, attainment8_average]
Index: []


In [65]:
#Investigate why some local authorities appear twice for the same year/status
duplicate_check = merged_df_clean[
    (merged_df_clean['lad_name'] == 'City of London') &
    (merged_df_clean['time_period'] == 202425) &
    (merged_df_clean['disadvantage_status'] == 'Total')
]

print(duplicate_check[['lad_name', 'time_period', 'disadvantage_status',
                        'breakdown_topic', 'breakdown', 'attainment8_average', 'pupil_count']])

            lad_name  time_period disadvantage_status breakdown_topic                     breakdown  attainment8_average  pupil_count
2660  City of London       202425               Total           Total                         Total                 63.7           16
2664  City of London       202425               Total      FSM status  Not known to be FSM eligible                 63.4           15


In [80]:
# Keep only rows where breakdown_topic is NOT 'FSM status'
# This removes FSM-based duplicate rows while correctly keeping the
# 'Total' breakdown_topic rows (which contain disadvantage_status = Total/Disadvantaged/Not disadvantaged)

print("BEFORE filtering breakdown_topic:", merged_df_clean.shape)

merged_df_clean = merged_df_clean[merged_df_clean['breakdown_topic'] != 'FSM status'].copy()

print("AFTER filtering breakdown_topic:", merged_df_clean.shape)

# Re-check for the City of London duplicate issue
check_again = merged_df_clean[
    (merged_df_clean['lad_name'] == 'City of London') &
    (merged_df_clean['time_period'] == 202425) &
    (merged_df_clean['disadvantage_status'] == 'Total')
]
print("\nRows for City of London 2024/25 Total after fix:", len(check_again))
print(check_again[['lad_name', 'breakdown_topic', 'breakdown', 'attainment8_average', 'pupil_count']])

BEFORE filtering breakdown_topic: (6377, 197)
AFTER filtering breakdown_topic: (6377, 197)

Rows for City of London 2024/25 Total after fix: 1
            lad_name breakdown_topic breakdown  attainment8_average  pupil_count
2660  City of London           Total     Total                 63.7           16


In [81]:
#  deduplicate on the combination of identifying columns
# Keep only one row per (lad_code, time_period, disadvantage_status)

print("BEFORE deduplication:", merged_df_clean.shape)

# Sort so we deterministically keep the 'Total' breakdown_topic row when there's a choice
merged_df_clean = merged_df_clean.sort_values(
    by=['lad_code', 'time_period', 'disadvantage_status', 'breakdown_topic']
)

# Drop duplicates on the key identifying columns
merged_df_clean = merged_df_clean.drop_duplicates(
    subset=['lad_code', 'time_period', 'disadvantage_status'],
    keep='first'
).copy()

print("AFTER deduplication:", merged_df_clean.shape)

# Verify Blackpool now has 1 row per status per year
verify = merged_df_clean[
    (merged_df_clean['lad_name'] == 'Blackpool') &
    (merged_df_clean['time_period'] == 202425)
]
print("\nBlackpool 2024/25 rows after fix:")
print(verify[['lad_name', 'time_period', 'disadvantage_status',
              'attainment8_average', 'pupil_count']])

BEFORE deduplication: (6377, 197)
AFTER deduplication: (6377, 197)

Blackpool 2024/25 rows after fix:
     lad_name  time_period            disadvantage_status  attainment8_average  pupil_count
81  Blackpool       202425                  Disadvantaged                 30.7          660
82  Blackpool       202425  Not known to be disadvantaged                 44.7          814
80  Blackpool       202425                          Total                 38.4         1474


In [82]:
# Re-save the de-duplicated data to SQLite
conn = sqlite3.connect('../outputs/education_deprivation.db')
merged_df_clean.to_sql('attainment_deprivation', conn, if_exists='replace', index=False)
print("Database re-saved with clean de-duplicated data")
print(f"Final rows stored: {len(merged_df_clean)}")

Database re-saved with clean de-duplicated data
Final rows stored: 6377


In [83]:
# Query: Most deprived local authorities in 2024/25, ranked by Attainment 8
# Now that duplicates are removed, each LA appears exactly once

conn = sqlite3.connect('../outputs/education_deprivation.db')

query = """
    SELECT lad_name, imd_rank, imd_score, attainment8_average
    FROM attainment_deprivation
    WHERE disadvantage_status = 'Total'
    AND time_period = 202425
    ORDER BY imd_rank ASC
    LIMIT 10
"""

most_deprived = pd.read_sql_query(query, conn)
print("Top 10 most deprived local authorities (2024/25):")
print(most_deprived)

conn.close()

Top 10 most deprived local authorities (2024/25):
               lad_name     imd_rank  imd_score  attainment8_average
0             Blackpool  7160.755319  43.527543                 38.4
1              Hastings  8162.094340  37.422887                 39.6
2            Manchester  8217.874576  37.962946                 45.0
3              Sandwell  8631.078947  33.432174                 43.5
4  Barking and Dagenham  8722.460870  30.274339                 48.8
5                Newham  8805.935135  31.353103                 51.3
6             Leicester  8876.075000  33.932390                 45.2
7            Birmingham  8954.295903  37.480082                 45.7
8               Hackney  9003.684564  31.771255                 49.9
9            Nottingham  9094.072626  35.047726                 43.5


### Section 3 Summary — Cleaning Outcomes

| Step | Before | After | Notes |
|---|---|---|---|
| Filter to LA level | 22,451 rows | 21,611 rows | Removed National/Regional rows |
| Remove unmatched LAs | 21,611 | 21,251 | 19 LAs lost due to boundary changes (1.7%) |
| Pupil count ≥ 5 filter | 21,251 | 21,220 | Isles of Scilly / City of London small cohorts (0.15%) |
| Remove FSM breakdown | 21,220 | 12,737 | Avoided double-counting from FSM breakdown |
| De-duplicate on (LA, year, status) | 12,737 | **6,377** | Final analysis-ready dataset |

The final cleaned dataset contains **6,377 rows** representing **~320 local authorities × 7 academic years × 3 disadvantage categories**, merged with deprivation indicators from IMD 2019 and IMD 2025. The data was also stored in a SQLite database (`outputs/education_deprivation.db`) to demonstrate SQL integration.

**Initial finding spotted in cleaning:** The 10 most deprived local authorities show striking variation in attainment — Blackpool (38.4) sits well below the national average while Newham (51.3) and Hackney (49.9) significantly outperform it, despite similar levels of deprivation. This previews the "London effect" that will be explored in Section 4.

## Section 4 — In-Depth Analysis with Statistical Insights

This section answers our four research questions through statistical analysis:

1. Do more deprived local authorities consistently produce lower Attainment 8 scores? *(Correlation analysis)*
2. Has the attainment gap between disadvantaged and non-disadvantaged pupils changed between 2018/19 and 2024/25? *(Gap analysis over time)*
3. Which local authorities perform above or below expectations relative to their deprivation level? *(Residual analysis)*
4. Can educational attainment be predicted using deprivation-related indicators? *(Predictive modelling — Section 6)*

In [85]:
# Research Question 1: Does deprivation correlate with Attainment 8?
# We use Pearson correlation — appropriate for two continuous variables

from scipy.stats import pearsonr

# Use the most recent year, Total population (not split by disadvantage)
latest_total = merged_df_clean[
    (merged_df_clean['time_period'] == 202425) &
    (merged_df_clean['disadvantage_status'] == 'Total')
].copy()

# Calculate correlation between IMD score and Attainment 8
correlation, p_value = pearsonr(latest_total['imd_score'],
                                 latest_total['attainment8_average'])

print(f"Pearson correlation coefficient: {correlation:.4f}")
print(f"P-value: {p_value:.6f}")

# Interpret the result
if p_value < 0.05:
    significance = "statistically significant"
else:
    significance = "not statistically significant"

if abs(correlation) > 0.7:
    strength = "strong"
elif abs(correlation) > 0.4:
    strength = "moderate"
else:
    strength = "weak"

direction = "negative" if correlation < 0 else "positive"

print(f"\nInterpretation: {strength} {direction} correlation, {significance} (p < 0.05)")
print(f"\nThis means: as deprivation score increases, Attainment 8 {'decreases' if correlation < 0 else 'increases'}.")

Pearson correlation coefficient: -0.6069
P-value: 0.000000

Interpretation: moderate negative correlation, statistically significant (p < 0.05)

This means: as deprivation score increases, Attainment 8 decreases.
